gnd/src/main/java/com/google/android/gnd/persistence/local/room/RoomLocalDataStore.java

```diff
@@ -83,51 +84,65 @@ public Completable applyAndEnqueue(FeatureMutation mutation) {
             list ->
                 stream(list)
                     .map(f -> FeatureEntity.toFeature(f, project))
-                    .collect(toImmutableSet()));
+                    .collect(toImmutableSet()))
+        .subscribeOn(Schedulers.io());
   }
 
   // TODO(#127): Decouple from Project and remove project from args.
   @Override
   public Maybe<Feature> getFeature(Project project, String featureId) {
-    return db.featureDao().findById(featureId).map(f -> FeatureEntity.toFeature(f, project));
+    return db.featureDao()
+        .findById(featureId)
+        .map(f -> FeatureEntity.toFeature(f, project))
+        .subscribeOn(Schedulers.io());
   }
 
   @Override
   public Maybe<Record> getRecord(Feature feature, String recordId) {
-    return db.recordDao().findById(recordId).map(record -> RecordEntity.toRecord(feature, record));
+    return db.recordDao()
+        .findById(recordId)
+        .map(record -> RecordEntity.toRecord(feature, record))
+        .subscribeOn(Schedulers.io());
   }
 
   @Override
-  public Flowable<ImmutableList<Record>> getRecordsOnceAndStream(Feature feature, String formId) {
+  public Single<ImmutableList<Record>> getRecords(Feature feature, String formId) {
     return db.recordDao()
-        .findByFeatureIdOnceAndStream(feature.getId(), formId)
+        .findByFeatureId(feature.getId(), formId)
         .map(
             list ->
                 stream(list)
                     .map(record -> RecordEntity.toRecord(feature, record))
-                    .collect(toImmutableList()));
+                    .collect(toImmutableList()))
+        .subscribeOn(Schedulers.io());
   }
 
   @Override
   public Flowable<ImmutableSet<Tile>> getTilesOnceAndStream() {
     return db.tileDao()
         .findAll()
-        .map(list -> stream(list).map(TileEntity::toTile).collect(toImmutableSet()));
+        .map(list -> stream(list).map(TileEntity::toTile).collect(toImmutableSet()))
+        .subscribeOn(Schedulers.io());
   }
 
   @Override
   public Single<ImmutableList<Mutation>> getPendingMutations(String featureId) {
     return db.featureMutationDao()
         .findByFeatureId(featureId)
-        .zipWith(db.recordMutationDao().findByFeatureId(featureId), this::mergeMutations);
+        .zipWith(db.recordMutationDao().findByFeatureId(featureId), this::mergeMutations)
+        .subscribeOn(Schedulers.io());
   }
 
   @Transaction
   @Override
   public Completable updateMutations(ImmutableList<Mutation> mutations) {
     return db.featureMutationDao()
         .updateAll(toFeatureMutationEntities(mutations))
-        .andThen(db.recordMutationDao().updateAll(toRecordMutationEntities(mutations)));
+        .andThen(
+            db.recordMutationDao()
+                .updateAll(toRecordMutationEntities(mutations))
+                .subscribeOn(Schedulers.io()))
```

Comment:
Do we know if this inner call to `subscribeOn` is necessary? Since `andThen` passes subscription to its argument, I'd assume it honors the scheduler too.

Reply:
Unfortunately it doesn't; the inner stream is created on a separate thread as with flatMap (I tested this).

Reply:
Dang! gtk thanks!